src/memo/memo.ml

```diff
@@ -704,16 +680,150 @@ module Call_stack = struct
+    let get_call_stack_tip () = get_call_stack () >>| List.hd_opt in
+    fun dep_node ->
+      let+ caller = get_call_stack_tip () in
+      match caller with
+      | None -> ()
+      | Some caller -> Stack_frame_with_state.add_dep caller ~dep_node
+end
+
+(* This module contains the essence of our cycle detection algorithm. Briefly,
+   the idea is as follows: whenever we are about to get blocked to wait for the
+   result of a computation that is currently running, we add the current call
+   stack to the cycle detection DAG. If this creates a cycle, we stop and report
+   a "dependency cycle" error; otherwise, we proceed with the blocking.
+
+   Below are some notes on how/why this algorithm works.
+
+   By "computation", we mean execution of a fiber. To share computation results
+   between multiple concurrent readers, we use [Once.t], which is a thin wrapper
+   around [Fiber.t] that executes the underlying fiber at most once. The first
+   reader "forces" the fiber and eventually gets the result, while other readers
+   subscribe to the future result of the computation via a [Fiber.Ivar.t]. This
+   blocks their execution and can lead to a deadlock if there is a dependency
+   cycle between different computations. To check for cycles *before* getting
+   blocked, a reader can use [Once.force_with_blocking_check].
+
+   One simple algorithm to check for cycles is to create a DAG node for every
+   computation, and add a DAG edge whenever a computation would like to read the
+   result of another unfinished computation. If adding an edge creates a cycle,
+   then we stop and report an error instead of getting blocked and deadlocked.
+   This algorithm is simple and it works, and Memo used it in the past. However,
+   the resulting DAG was often large, and so cycle detection was taking ~35% of
+   incremental zero rebuilds. As an optimisation, we developed another algorithm
+   described below.
+
+   The DAG produced by the above algorithm can contain many uninteresting nodes
+   and edges. For example, every node will have at least one incoming edge that
+   is added when the corresponding computation was initially "forced". But these
+   "forcing edges" cannot cause deadlocks by themselves; in fact, if there is no
+   parallelism, there may be no "blocking edges" at all, and the above algorithm
+   would add all those forcing edges for nothing.
+
+   Here is an optimisation idea. Since blocking edges are the real cause of
+   deadlocks, we will focus our attention on them: when we hit a blocking edge,
+   we will add it to the DAG *along with the path that led us to it*. This path
+   is readily available to us in the form of the call stack.
+
+   Here is a proof sketch that this algorithm finds all possible cycles. We are
+   going to make use of the following three observations:
+
+   (1) If there is a cycle, it must contain at least one blocking edge.
+
+   (2) Every path that we are going to add to the DAG will contain a sequence of
```

Comment:
I guess it's important that the paths are "rooted" at the root, otherwise when tracing back the forcing edges we'd have to deal with the possibility that we "run out" of stack.

I don't know if this is the level of detail you deliberately omitted or still want to include, though.

Reply:
I don't think having a single "root" is actually important. Is it? Silently assuming a finite call stack seems fine to me :) Let me know if I'm missing anything.

Reply:
Perhaps by "running out of stack" you mean the stack will end before we manage to add all forcing edges that lie on the cycle? In this case, I think this particular cycle has not yet been reached.

Reply:
Yeah, agreed, we don't need a unique root and I agree there's no issue with the algorithm, merely that there's a detail that's missing from the proof. 

We do need the full stack, not an "arbitrarily truncated" one. I think the bullet point (2) technically allows a truncated one.